cloudsync/smartsync.py

```diff
@@ -0,0 +1,374 @@
+        self._resolve_conflict = resolve_conflict
+        self.tempdir = tempfile.mkdtemp(suffix=".cloudsync")
+        self.__nmgr = notification_manager
+        self._root_oids: List[str] = list(root_oids) if root_oids else [None, None]
+        self._root_paths: List[str] = list(root_paths) if root_paths else [None, None]
+        if not sleep:
+            # these are the event sleeps, but really we need more info than this
+            sleep = (self.providers[LOCAL].default_sleep, self.providers[REMOTE].default_sleep)
+
+        self.sleep = sleep
+
+        ####
+
+        max_sleep = max(sleep)                    # on sync fail, use the worst time for backoff
+        self.aging = max_sleep / 5                # how long before even trying to sync
+        self.min_backoff = max_sleep / 10.0       # event sleep of 15 seconds == 1.5 second backoff on failures
+        self.max_backoff = max_sleep * 10.0       # escalating up to a 3 minute wait time
+        self.mult_backoff = 2
+        assert len(self.providers) == 2
+    def pre_sync(self, sync: SyncEntry) -> bool:  # pylint: disable=too-many-branches
+        finished = super().pre_sync(sync)
+        local_file = sync[LOCAL].oid and self.providers[LOCAL].exists_oid(sync[LOCAL].oid)
+        if not finished:
+            finished = not (local_file or sync in self.state.requestset or sync[REMOTE].otype == DIRECTORY)
+        return finished
+
+    def get_parent_conflicts(self, sync: SyncEntry, changed) -> List[SyncEntry]:
+        """Returns list of parent conflicts."""
+        done = False
+        pcs = []
+        new_pc = sync
+        while not done:
+            pc = self._get_parent_conflict(new_pc, changed)
+            # if pc:  # TODO: test if this is necessary
+            #     pc.get_latest()
+            if pc and pc[REMOTE].path != new_pc[REMOTE].path:
+                pcs.append(pc)
+                new_pc = pc
+            else:
+                done = True
+        pcs.reverse()
+        return pcs
+
+
+class SmartSyncState(SyncState):
+    """Enhances the syncstate to support smart syncing."""
+    def __init__(self,
+                 providers: Tuple['Provider', 'Provider'],
+                 storage: Optional['Storage'] = None,
+                 tag: Optional[str] = None,
+                 shuffle: bool = False,
+                 prioritize: Callable[[int, str], int] = None):
+        self.requestset: Set[SyncEntry] = set()
+        self.excludeset: Set[SyncEntry] = set()
+        self._callbacks: List[Callable] = list()
+        super().__init__(providers, storage, tag, shuffle, prioritize)
+
+    def register_auto_sync_callback(self, callback):
+        self._callbacks.append(callback)
+
+    def _smart_sync_ent(self, ent):
+        if ent[LOCAL].path and not self.providers[LOCAL].exists_path(ent[LOCAL].path):
+            ent[LOCAL].clear()
+            ent[REMOTE].sync_path = None
```

Comment:
Why do we clear sync_path and sync_hash?

Reply:
Because if the local file doesn't exist, then the file isn't synced, and if the remote has sync info then that info is stale at that point and the stale info would prevent the file from syncing down

Reply:
Why is the info stale? And why would it prevent the file from syncing down? I feel like I'm missing something

Reply:
If the file does not exist locally, then the remote file is not synced, so how can it be said that a sync_path and sync_hash that are not None on the remote side are accurate?

Reply:
And if `path==sync_path` and the `hash==sync_hash` on the remote side, why would the sync manager believe that the remote side `needs_sync`?

Reply:
Hmm I see. Okay I'm convinced